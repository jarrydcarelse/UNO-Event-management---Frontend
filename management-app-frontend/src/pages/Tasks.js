import React, { useState, useEffect } from 'react';
import Navbar from '../components/Navbar';
import axios from 'axios';
import '../tasks/Tasks.css';

// Update API_BASE to use the deployed backend URL
const API_BASE = 'https://eventify-backend-kgtm.onrender.com';

// Add axios default configuration
axios.defaults.baseURL = API_BASE;
axios.defaults.headers.common['Access-Control-Allow-Origin'] = '*';
axios.defaults.headers.common['Access-Control-Allow-Methods'] = 'GET,PUT,POST,DELETE,PATCH,OPTIONS';
axios.defaults.headers.common['Access-Control-Allow-Headers'] = 'Content-Type, Authorization';

export default function Tasks() {
  const [sidebarOpen, setSidebarOpen] = useState(true);
  const [tasks, setTasks] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');

  // Fetch tasks on component mount
  useEffect(() => {
    const fetchTasks = async () => {
      try {
        const token = localStorage.getItem('token');
        if (!token) {
          navigate('/login');
          return;
        }

        const response = await axios.get('/api/eventtasks', {
          headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
            'Accept': 'application/json'
          }
        });

        // Transform the tasks data to match our display format
        const formattedTasks = response.data.map(task => ({
          id: task.id,
          title: task.title,
          description: task.description,
          dueDate: task.dueDate,
          priority: task.priority,
          priorityClass: task.priority === 'High' ? 'red' : task.priority === 'Medium' ? 'yellow' : 'green',
          assignedTo: task.assignedToEmail,
          completed: task.completed,
          budget: task.budget,
          eventId: task.eventId
        }));

        setTasks(formattedTasks);
        setLoading(false);
      } catch (error) {
        console.error('Error fetching tasks:', error);
        setError('Failed to load tasks. Please try again later.');
        setLoading(false);
      }
    };

    fetchTasks();
  }, []);

  const inProgressTasks = tasks.filter(task => !task.completed);
  const completedTasks = tasks.filter(task => task.completed);
  const totalTasks = tasks.length;
  const completedCount = completedTasks.length;
  const progressPercentage = totalTasks > 0 ? Math.round((completedCount / totalTasks) * 100) : 0;

  if (loading) {
    return (
      <div className="tasks-layout">
        <Navbar isOpen={sidebarOpen} setIsOpen={setSidebarOpen} />
        <div className={`tasks-page${sidebarOpen ? '' : ' collapsed'}`}>
          <div className="tasks-header">
            <h1>Tasks Overview</h1>
          </div>
          <div className="tasks-section">
            <p>Loading tasks...</p>
          </div>
        </div>
      </div>
    );
  }

  if (error) {
    return (
      <div className="tasks-layout">
        <Navbar isOpen={sidebarOpen} setIsOpen={setSidebarOpen} />
        <div className={`tasks-page${sidebarOpen ? '' : ' collapsed'}`}>
          <div className="tasks-header">
            <h1>Tasks Overview</h1>
          </div>
          <div className="tasks-section">
            <p className="error-message">{error}</p>
          </div>
        </div>
      </div>
    );
  }

  return (
    <div className="tasks-layout">
      <Navbar isOpen={sidebarOpen} setIsOpen={setSidebarOpen} />
      <div className={`tasks-page${sidebarOpen ? '' : ' collapsed'}`}>
        <div className="tasks-header">
          <h1>Tasks Overview</h1>
        </div>

        {/* Progress Bar Section */}
        <div className="tasks-progress-section">
          <div className="progress-stats">
            <div className="progress-stat">
              <span className="stat-label">Total Tasks</span>
              <span className="stat-value">{totalTasks}</span>
            </div>
            <div className="progress-stat">
              <span className="stat-label">Completed</span>
              <span className="stat-value">{completedCount}</span>
            </div>
            <div className="progress-stat">
              <span className="stat-label">In Progress</span>
              <span className="stat-value">{inProgressTasks.length}</span>
            </div>
          </div>
          <div className="progress-bar-container">
            <div className="progress-bar-bg">
              <div 
                className="progress-bar-fill"
                style={{ 
                  width: `${progressPercentage}%`,
                  backgroundColor: progressPercentage === 100 ? '#FF8FAB' : progressPercentage > 50 ? '#FF8FAB' : '#FF8FAB'
                }}
              />
            </div>
            <span className="progress-percentage">{progressPercentage}%</span>
          </div>
        </div>

        {/* In Progress Tasks Section */}
        <div className="tasks-section">
          <h2>In Progress</h2>
          <div className="tasks-grid">
            {inProgressTasks.length === 0 ? (
              <p className="no-tasks">No tasks in progress</p>
            ) : (
              inProgressTasks.map(task => (
                <div key={task.id} className="task-card">
                  <div className="task-header">
                    <h3 className="task-name">{task.title}</h3>
                    <span className={`priority-badge ${task.priority.toLowerCase()}`}>
                      {task.priority}
                    </span>
                  </div>
                  <hr className="inner-divider" />
                  
                  <p className="task-desc">
                    Description: <strong>{task.description}</strong>
                  </p>
                  <hr className="inner-divider" />
                  
                  <div className="task-meta">
                    <p>
                      Assigned To: <strong>{task.assignedTo}</strong>
                    </p>
                    <p>
                      Due Date: <strong>{new Date(task.dueDate).toLocaleDateString()}</strong>
                    </p>
                    <p>
                      Budget: <strong>{task.budget}</strong>
                    </p>
                  </div>
                </div>
              ))
            )}
          </div>
        </div>

        {/* Completed Tasks Section */}
        <div className="tasks-section">
          <h2>Completed</h2>
          <div className="tasks-grid">
            {completedTasks.length === 0 ? (
              <p className="no-tasks">No completed tasks</p>
            ) : (
              completedTasks.map(task => (
                <div key={task.id} className="task-card completed">
                  <div className="task-header">
                    <h3 className="task-name">{task.title}</h3>
                    <span className={`priority-badge ${task.priority.toLowerCase()}`}>
                      {task.priority}
                    </span>
                  </div>
                  <hr className="inner-divider" />
                  
                  <p className="task-desc">
                    Description: <strong>{task.description}</strong>
                  </p>
                  <hr className="inner-divider" />
                  
                  <div className="task-meta">
                    <p>
                      Assigned To: <strong>{task.assignedTo}</strong>
                    </p>
                    <p>
                      Due Date: <strong>{new Date(task.dueDate).toLocaleDateString()}</strong>
                    </p>
                    <p>
                      Budget: <strong>{task.budget}</strong>
                    </p>
                  </div>
                </div>
              ))
            )}
          </div>
        </div>
      </div>
    </div>
  );
} 