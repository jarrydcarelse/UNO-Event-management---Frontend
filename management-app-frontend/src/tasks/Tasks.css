body {
  font-family: 'Manrope', sans-serif;
  color: #111827;
  background: #f2f2f7;
  margin: 0;
}

.tasks-layout {
  display: flex;
  height: 100vh;
}

.tasks-page {
  flex: 1;
  padding: 32px;
  margin-left: 250px;
  box-sizing: border-box;
  transition: margin-left 0.3s ease;
  width: 100%;
  background: #f2f2f7;
}

.tasks-page.collapsed {
  margin-left: 100px;
}

.tasks-header {
  background: #fff;
  border-radius: 16px;
  padding: 26px 32px 18px 32px;
  margin-bottom: 18px;
  box-shadow: 0 2px 6px rgba(0,0,0,0.05);
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.tasks-header h1 {
  font-size: 22px;
  font-weight: 700;
  margin: 0;
  color: #232336;
}

.add-task-btn {
  background: #FF5F8A;
  color: #fff;
  font-weight: 700;
  border: none;
  border-radius: 999px;
  font-size: 16px;
  padding: 10px 22px;
  box-shadow: 0 2px 8px rgba(255,95,138,0.15);
  display: flex;
  align-items: center;
  gap: 8px;
  transition: background 0.2s;
  cursor: pointer;
}

.add-task-btn:hover {
  background: #e94a73;
}

.tasks-section {
  background: #fff;
  border-radius: 16px;
  padding: 28px 32px 32px 32px;
  box-shadow: 0 2px 6px rgba(0,0,0,0.05);
  margin-bottom: 40px;
}

.tasks-section h2 {
  font-size: 17px;
  font-weight: 700;
  color: #232336;
  margin: 0 0 18px 0;
}

.tasks-grid {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(320px, 1fr));
  align-items: start;
  gap: 22px;
  justify-content: left;
}

.task-card {
  background: #f2f2f7;
  border-radius: 12px;
  padding: 22px 18px 18px 18px;
  min-width: 220px;
  max-width: 360px;
  box-shadow: none;
  display: flex;
  flex-direction: column;
  gap: 13px;
  position: relative;
  border: 2px solid transparent;
  transition: border 0.2s;
}

.task-card.completed {
  border: 2px solid #b6f2c0;
  background: #e8fff0;
}

.task-header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  font-weight: 600;
  font-size: 16px;
  margin-bottom: 4px;
}

.task-header h3 {
  margin: 0;
  font-size: 16px;
  font-weight: 600;
  color: #232336;
}

.priority-badge {
  padding: 4px 12px;
  border-radius: 12px;
  font-size: 12px;
  font-weight: 600;
}

.priority-badge.low {
  background: #E3F2FD;
  color: #1976D2;
}

.priority-badge.medium {
  background: #FFF3E0;
  color: #F57C00;
}

.priority-badge.high {
  background: #FFEBEE;
  color: #D32F2F;
}

.task-description {
  margin: 8px 0;
  font-size: 14px;
  color: #666;
  line-height: 1.4;
}

.task-meta {
  display: flex;
  flex-wrap: wrap;
  gap: 12px;
  font-size: 13px;
  color: #666;
}

.task-meta span {
  display: flex;
  align-items: center;
  gap: 4px;
}

.task-actions {
  display: flex;
  gap: 13px;
  margin-top: auto;
  flex-wrap: wrap;
}

.task-actions button {
  display: flex;
  align-items: center;
  gap: 5px;
  border: none;
  background: none;
  color: #FF5F8A;
  font-size: 15px;
  font-weight: 600;
  cursor: pointer;
  padding: 0;
  border-radius: 7px;
  transition: background 0.2s;
}

.task-actions button:hover {
  background: #fce6ef;
}

.task-actions button svg {
  font-size: 16px;
}

/* Modal Styles */
.modal-overlay {
  position: fixed;
  z-index: 9999;
  left: 0;
  top: 0;
  width: 100vw;
  height: 100vh;
  background: rgba(34, 34, 54, 0.24);
  display: flex;
  align-items: center;
  justify-content: center;
  animation: fadeIn 0.18s;
}

@keyframes fadeIn {
  from { opacity: 0; }
  to { opacity: 1; }
}

.modal {
  background: #fff;
  border-radius: 20px;
  padding: 34px 32px;
  min-width: 320px;
  max-width: 95vw;
  animation: popIn 0.2s;
}

@keyframes popIn {
  from { transform: scale(0.93); opacity: 0.6; }
  to { transform: scale(1); opacity: 1; }
}

.modal-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 24px;
}

.modal-header h2 {
  margin: 0;
  font-size: 24px;
  font-weight: 700;
  color: #232336;
}

.close-btn {
  background: none;
  border: none;
  color: #666;
  cursor: pointer;
  padding: 8px;
  transition: color 0.2s;
}

.close-btn:hover {
  color: #d61a51;
}

.form-group {
  margin-bottom: 20px;
}

.form-group label {
  display: block;
  margin-bottom: 8px;
  font-weight: 500;
  color: #232336;
}

.form-group input,
.form-group textarea,
.form-group select {
  width: 100%;
  padding: 10px;
  border: 1px solid #ddd;
  border-radius: 6px;
  font-size: 14px;
  transition: border-color 0.2s;
}

.form-group input:focus,
.form-group textarea:focus,
.form-group select:focus {
  outline: none;
  border-color: #FF5F8A;
}

.form-group textarea {
  min-height: 100px;
  resize: vertical;
}

.modal-actions {
  display: flex;
  justify-content: flex-end;
  margin-top: 24px;
}

.save-btn {
  background: #FF5F8A;
  color: white;
  border: none;
  border-radius: 6px;
  padding: 12px 24px;
  font-size: 16px;
  font-weight: 600;
  cursor: pointer;
  transition: all 0.2s ease;
}

.save-btn:hover {
  background: #e94a73;
}

@media (max-width: 900px) {
  .tasks-header,
  .tasks-section {
    padding: 20px;
  }
  
  .tasks-page {
    margin-left: 80px;
    padding: 12px;
  }
  
  .tasks-grid {
    grid-template-columns: 1fr;
  }
} 