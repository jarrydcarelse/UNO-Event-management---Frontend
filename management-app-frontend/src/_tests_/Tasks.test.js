import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import Tasks from '../pages/Tasks';
import axios from 'axios';

jest.mock('axios');
jest.mock('../components/Navbar', () => () => <div data-testid="navbar" />);
jest.mock('../components/LoadingSpinner', () => () => <div data-testid="spinner" />);

describe('Tasks Component', () => {
  beforeEach(() => {
    localStorage.setItem('token', 'fake-token');
  });

  afterEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

  it('renders loading state initially', async () => {
    axios.get.mockImplementation(() => new Promise(() => {})); // simulate loading
    render(<Tasks />);
    expect(screen.getByTestId('spinner')).toBeInTheDocument();
  });

  it('renders tasks after fetch', async () => {
    axios.get.mockResolvedValue({
      data: [
        {
          id: 1,
          title: 'Task 1',
          description: 'Do something',
          dueDate: new Date().toISOString(),
          priority: 'High',
          assignedToEmail: 'user@example.com',
          completed: false,
          budget: 500,
          eventId: 123,
        },
        {
          id: 2,
          title: 'Task 2',
          description: 'Do another thing',
          dueDate: new Date().toISOString(),
          priority: 'Low',
          assignedToEmail: 'another@example.com',
          completed: true,
          budget: 1000,
          eventId: 123,
        },
      ],
    });

    const { container } = render(<Tasks />);
    await waitFor(() => {
      expect(screen.getByText('Tasks Overview')).toBeInTheDocument();
    });

    expect(screen.getByText('Task 1')).toBeInTheDocument();
    expect(screen.getByText('Task 2')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('shows error message on fetch fail', async () => {
    axios.get.mockRejectedValue(new Error('Network error'));

    render(<Tasks />);
    await waitFor(() => {
      expect(screen.getByText(/failed to load tasks/i)).toBeInTheDocument();
    });
  });
});
