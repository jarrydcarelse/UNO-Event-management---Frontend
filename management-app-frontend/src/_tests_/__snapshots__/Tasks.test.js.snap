// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks Component renders tasks after fetch 1`] = `
<div>
  <div
    class="tasks-layout"
  >
    <div
      data-testid="navbar"
    />
    <div
      class="tasks-page"
    >
      <div
        class="tasks-header"
      >
        <h1>
          Tasks Overview
        </h1>
      </div>
      <div
        class="tasks-progress-section"
      >
        <div
          class="progress-stats"
        >
          <div
            class="progress-stat"
          >
            <span
              class="stat-label"
            >
              Total Tasks
            </span>
            <span
              class="stat-value"
            >
              2
            </span>
          </div>
          <div
            class="progress-stat"
          >
            <span
              class="stat-label"
            >
              Completed
            </span>
            <span
              class="stat-value"
            >
              1
            </span>
          </div>
          <div
            class="progress-stat"
          >
            <span
              class="stat-label"
            >
              In Progress
            </span>
            <span
              class="stat-value"
            >
              1
            </span>
          </div>
        </div>
        <div
          class="progress-bar-container"
        >
          <div
            class="progress-bar-bg"
          >
            <div
              class="progress-bar-fill"
              style="width: 50%; background-color: rgb(255, 143, 171);"
            />
          </div>
          <span
            class="progress-percentage"
          >
            50
            %
          </span>
        </div>
      </div>
      <div
        class="tasks-section"
      >
        <h2>
          In Progress
        </h2>
        <div
          class="tasks-grid"
        >
          <div
            class="task-card"
          >
            <div
              class="task-header"
            >
              <h3
                class="task-name"
              >
                Task 1
              </h3>
              <span
                class="priority-badge high"
              >
                High
              </span>
            </div>
            <hr
              class="inner-divider"
            />
            <p
              class="task-desc"
            >
              Description: 
              <strong>
                Do something
              </strong>
            </p>
            <hr
              class="inner-divider"
            />
            <div
              class="task-meta"
            >
              <p>
                Assigned To: 
                <strong>
                  user@example.com
                </strong>
              </p>
              <p>
                Due Date: 
                <strong>
                  2025/08/05
                </strong>
              </p>
              <p>
                Budget: 
                <strong>
                  500
                </strong>
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="tasks-section"
      >
        <h2>
          Completed
        </h2>
        <div
          class="tasks-grid"
        >
          <div
            class="task-card completed"
          >
            <div
              class="task-header"
            >
              <h3
                class="task-name"
              >
                Task 2
              </h3>
              <span
                class="priority-badge low"
              >
                Low
              </span>
            </div>
            <hr
              class="inner-divider"
            />
            <p
              class="task-desc"
            >
              Description: 
              <strong>
                Do another thing
              </strong>
            </p>
            <hr
              class="inner-divider"
            />
            <div
              class="task-meta"
            >
              <p>
                Assigned To: 
                <strong>
                  another@example.com
                </strong>
              </p>
              <p>
                Due Date: 
                <strong>
                  2025/08/05
                </strong>
              </p>
              <p>
                Budget: 
                <strong>
                  1000
                </strong>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
